### ISSUE TITLE
Is there a plan to add a reverse tokenization pipeline?

### ISSUE THREAD

**Issue by vincentzlt** [huggingface/tokenizers#563]
Machine translation tasks evaluate based on the original text. So it is important to have all the encoded ids transform safely to the original ids. The current decoders are only sufficient in decoding the ids to the text after pre_tokenizers. Is it possible to add some functionality that decodes to the original text level?

**Comment by n1t0** [comment]
Hi @vincentzlt 

The ability to revert back to the original text depends on the tokenizer you are building. When some destructive normalization is applied, there is no way to revert back to the original text using the IDs only.

You can maybe build a tokenizer using the byte-level pre-tokenizers/decoders that doesn't normalize too much to achieve this:
```python
from tokenizers import Tokenizer, models, normalizers, pre_tokenizers, decoders

tokenizer = Tokenizer(models.Unigram())
# or
tokenizer = Tokenizer(models.BPE())

tokenizer.normalizer = normalizers.NFKC()
tokenizer.pre_tokenizer = pre_tokenizers.ByteLevel()
tokenizer.decoder = decoders.ByteLevel()

trainer = tokenizer.model.get_trainer()
trainer.vocab_size = 20000 # Set the target vocabulary size
trainer.special_tokens = ["<pad>", "<cls>", "<sep>"] # Set some special tokens if needed
trainer.initial_alphabet = pre_tokenizers.ByteLevel.alphabet() # Use ByteLevel alphabet

tokenizer.train_from_iterator(dataset, trainer=trainer)
```

(This example will only work with a the version `0.10.0rc1` or higher, you can install it using `pip install tokenizers --pre`)

**Comment by n1t0** [comment]
Closing this, feel free to reopen if needed!

**Comment by kitkhai** [comment]
Hi @n1t0 ,

In the [docs](https://huggingface.co/docs/tokenizers/components#normalizers:~:text=The%20specificity%20of%20tokenizers%20is%20that%20we%20keep%20track%20of%20the%20alignment%20while%20normalizing.%20This%20is%20essential%20to%20allow%20mapping%20from%20the%20generated%20tokens%20back%20to%20the%20input%20text), it was stated that "**The specificity of tokenizers is that we keep track of the alignment while normalizing. This is essential to allow mapping from the generated tokens back to the input text.**"

So how do I access the mapping such that I can recreate the input text from the generated tokens?

For example, a comma is normalized into another comma so how do I retrieve the original pre-normalized comma?
```
checkpoint = "facebook/nllb-200-distilled-600M"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

pre_normalized_comma = "，"
print(tokenizer.tokenize(pre_normalized_comma))                          # ['▁,']
print(tokenizer.encode(pre_normalized_comma))                            # [256047, 146, 2]
print(tokenizer.convert_tokens_to_ids(pre_normalized_comma))             # 3
print(tokenizer.convert_ids_to_tokens(3))                                # <unk>
```

**Comment by ArthurZucker** [comment]
You are using `tokenizers` through the `transformers` API. To get the mapping and everything else you need to run: 
```python 
checkpoint = "facebook/nllb-200-distilled-600M"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

pre_normalized_comma = "，"
print(tokenizer._tokenizer.encode(pre_normalized_comma))
Encoding(num_tokens=3, attributes=[ids, type_ids, tokens, offsets, attention_mask, special_tokens_mask, overflowing])
```

**Comment by kitkhai** [comment]
Hi @ArthurZucker,

Thanks for the reply

I get that the encoding object will give me the offsets and everything else which does allow mapping from the generated tokens back to the input text.
But is there a mapping specifically between the normalized and pre-normalized tokens?

**Comment by ArthurZucker** [comment]
I think the `offsets` should give this, and if you want only the ones relates to normalization. I would probably set an identity normalizer for example!
